UIE-147 Environments Page - decouple permissions

msilva-broad opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/UIE-147

<!-- ### Dependencies --->
<!-- Include any dependent tickets and describe the relationship. Include any relevant Jira tickets. --->

## Summary of changes:
refactored permissions related logic to be composable instead of hard-wired inside Environments component.

### What
- introduced EnvironmentsPermissionsProvider type and implementation which is now a required prop of Environments component.  This decouples permissions logic so that it can be more easily composed by AoU or other consumers.
- improved related types for code in the path of permissions logic (previously using current user email directly)
- improved clarity around raw disk api data types.
- fixed bug where renderErrorApps was using getDisplayRuntimeStatus helper, instead of getAppStatusForDisplay.
- also created EnvironmentsPage sub-directory for new bits, but will hold off on moving existing EnvironmentsPage bits into it.
---
- PR feedback - renamed permission provider and types.
- changed permission methods to accept narrow IHaveCreator argument, tweaked related bits to use new type.

### Why
- part of enabling code sharing with AoU

### Testing strategy
- smoke/regression test, no functional changes.

<!-- ### Visual Aids -->
<!-- https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac -->

msilva-broad committed:
UIE-147 Decouple permissions - Environments page
- introduced EnvironmentsPermissionsProvider type and implementation which is now a required prop of Environments component.  This decouples permissions logic so that it can be more easily composed by AoU or other consumers.
- improved related types for code in the path of permissions logic (previously using current user email directly)
- improved clarity around raw disk api data types.
- fixed bug where renderErrorApps was using getDisplayRuntimeStatus helper, instead of getAppStatusForDisplay.
- also created EnvironmentsPage sub-directory for new bits, but will hold off on moving existing EnvironmentsPage bits into it.

msilva-broad committed:
UIE-147 Decouple permissions - Environments page
- small rebase fix

jdcanas commented on src/analysis/Environments/Environments.ts:
```diff
@@ -207,6 +210,10 @@ export function PauseButton(props: PauseButtonProps): ReactNode {
 type LeoAppProviderNeeds = Pick<LeoAppProvider, 'listWithoutProject' | 'get' | 'pause' | 'delete'>;
 type LeoRuntimeProviderNeeds = Pick<LeoRuntimeProvider, 'list' | 'stop' | 'delete'>;
 type LeoDiskProviderNeeds = Pick<LeoDiskProvider, 'list' | 'delete'>;
+export interface EnvironmentPermissionsProvider {
```
Can we call this ResourcePermissionProvider? That's more in line with what we call functions that can handle `(resource: PersistenDisk | Runtime | App)` over some common field like `auditInfo`

msilva-broad replied:
yeah... I think I like that.... gonna think about the Resource arg tho... might be better to define a narrow interface IHavePermissions.... which all three implement.

msilva-broad replied:
went with LeoResourcePermissionsProvider and IHaveCreator

jdcanas commented on src/pages/EnvironmentsPage/environmentsPermissions.ts:
```diff
@@ -0,0 +1,26 @@
+      const canDelete = disk.auditInfo.creator === currentUserEmail;
+      return canDelete;
+    },
+    canPauseResource: (resource) => {
```
It could be worth typing this here as `App | PersistentDisk | Runtime`. Or making a `Resource` type that is the union of the three

msilva-broad replied:
see above.... thinking I might go with narrow interface IHavePermissions, or better yet... IHaveOwner maybe which all three implement.

msilva-broad replied:
went with narrow IHaveCreator arg interface

jdcanas commented on src/analysis/Environments/Environments.ts:
```diff
@@ -781,11 +789,12 @@ export const Environments = (props: EnvironmentsProps): ReactNode => {
                 cellRenderer: ({ rowIndex }) => {
-                  const { status: diskStatus, googleProject, workspace, creator } = filteredDisks[rowIndex];
+                  const rowDisk = filteredDisks[rowIndex];
+                  const { status: diskStatus, cloudContext, workspace } = rowDisk;
```
can we pull the creator off the audit info here? I believe the downstream component has that as optional (in the logic, not with sensible types to indicate such), so currently it just isn't displayed. The info is there for all disks though. It's just not being extracted properly.

msilva-broad replied:
I decided to have the implementation of permissions.canDeleteDisk be where poking into audit info happens.  So that that concern is removed from the component here.

jdcanas replied:
Ah sorry, I thought this was for a different column at first (the details view column), which displays this field properly.

jdcanas commented on src/libs/ajax/leonardo/models/disk-models.ts:
```diff
@@ -31,37 +32,25 @@ export const diskStatuses: { [label: string]: DiskStatus } = {
-  diskType: GoogleDiskType;
-  blockSize: number;
-  labels: LeoResourceLabels;
+export interface RawGetDiskItem extends RawListDiskItem {
```
I don't know if I like these inheriting from one another. The fact that List is a subset of Get is a coincidence, and not always the case. I believe for both apps and runtimes they've diverged, the motivation being the trade off between a fast query with as much info as possibly and a more costly join on the backend for certain fields. I can see the argument for keeping it as such until a time they do diverge since it makes logical since for one to be seen as the 'details' view though and don't feel too strongly.

msilva-broad replied:
i'm for changing it later if needed, but for now I think it adds clarity.

nawatts replied:
Agreed. I'd vote against linking two things for convenience when they're not conceptually related.

msilva-broad replied:
I think they are related in that the more detailed one just adds a small number of additional fields.  Using extends makes it clear which fields are added.  But if folks feel strongly, I can switch it back to non-extended pair of types.

evrii commented on src/pages/EnvironmentsPage/environmentsPermissions.ts:
```diff
@@ -0,0 +1,26 @@
+};
+
+const currentUserEmailGetter = (): string => {
+  return getTerraUser().email!;
```
It looks like this relies on Terra state, will this need to be abstracted away.

Is this one of the "Terra-specific" dependencies you mentioned on Slack?

evrii replied:
Never mind. I see now that this is the provider used by Terra where AoU would provide its own.

msilva-broad replied:
correct :)

jdcanas commented on src/pages/EnvironmentsPage/environmentsPermissions.test.ts:
```diff
@@ -0,0 +1,78 @@
+import { DeepPartial } from '@terra-ui-packages/core-utils';
+import { asMockedFn } from '@terra-ui-packages/test-utils';
```
cc @LizBaldo @lucymcnatt @cindy-broadinstitute @mlilynolting, this is particularly relevant as we recently discussed the disconnect between the UI enforcing IAM logic form the resource, and the backend knowing the information. I think this implementation is a great step in the right direction to that end, as it at the abstracts the permission logic away from the visual component and provides verbs like `canIPauseThisDisk` which is all the environments page cares about. Down the line, these verbs could leverage IAM if necessary for the business logic without affecting large areas of the code.

jdcanas reviewed (approved):
Few nits here and there, but validate the behavior of delete/pause on the cloud environments page, and functionally it looks good.

msilva-broad committed:
UIE-147 Decouple permissions - Environments page
- PR feedback - renamed permission provider and types.
- changed permission methods to accept narrow IHaveCreator argument, tweaked related bits to use new type.

msilva-broad committed:
UIE-147 Decouple permissions - Environments page
- minor fix

nawatts commented on src/analysis/Environments/Environments.ts:
```diff
@@ -384,7 +391,7 @@ export const Environments = (props: EnvironmentsProps): ReactNode => {
     [sort.direction],
     runtimes
-  );
+  ) as unknown as typeof runtimes;
```
An awful lot of casts in this file...

msilva-broad replied:
yes, because lodash/fp is horrible with type flow...

nawatts replied:
My experience has been that an issue with the output's type is often caused by an issue with the input's type.

msilva-broad replied:
agreed, I saw that input can be undefined, but that wasn't the only issue in some cases like here.... at any rate, I'd love to rip this data conversion stuff out to a helper where it can be more focused.

msilva-broad replied:
I'd like to rework all this data transformation to be a more focused helper to both remove bloat from this component, and to make it less wonky on the types.  But that's work for another time...

nawatts commented on src/libs/ajax/leonardo/models/core-models.ts:
```diff
@@ -10,6 +10,10 @@ export interface AuditInfo {
   dateAccessed: string;
 }
 
+export interface IHaveCreator {
```
Really dislike this name. Maybe `HasCreator`?

nawatts replied:
Actually... why do we even need this type? Environments passes the disk or runtime to the permissions functions, not a slice of those objects.

msilva-broad replied:
it's to narrow the required object shape so that any of the three can work (App, Runtime, Disk)

And sure, HasCreator is good.

nawatts replied:
Why create a narrower type vs using a union type to more precisely reflect what's passed?

msilva-broad replied:
following the I in SOLID, favor smaller interfaces over larger ones.  The narrower interface makes the permission provider more flexible since it can take anything that has thing.auditInfo.creator.... if needs change later, we can broaden the type.

nawatts replied:
That makes sense for inputs, but not really for callback types, where it's not the consumer that's providing something to satisfy the interface.

And if the type is going to limit consumers to only using the creator field instead of the object actually being passed, then why pass the full object at all vs just the creator email?

msilva-broad replied:
hmmm.... I get your logic.... How about I tweak it so that HasCreator is typed { creator: string } and the component can pass in resource.auditInfo?

nawatts replied:
Sure, though I'd still claim that if the component is going to pass in `resource.auditInfo`, the callback should be typed as taking an `AuditInfo` object.

nawatts replied:
But also, the more I think about it, the more I think this might be the wrong abstraction... The component doesn't need to ask the app to determine if the user has permission to access a resource. It _knows_ how to determine that. It only needs to know the user's email address. As-is, each app has to duplicate the same logic in the permissions provider.

msilva-broad replied:
yes, but I think there is value in separating the concern completely... this has the benefit of later centralizing and managing change in the newly isolated provided versus hunting down usage in various components.

Justin's earlier comment seems to agree with this.

msilva-broad replied:
ok, I tried out having it take in just the audit object with only creator field specified, but don't like how it feels... forces consumer to be more verbose... permissions.canDeleteDisk(disk.auditInfo) feels clunky compared to just permissions.canDeleteDisk(disk)...

So I think I will go back to the earlier suggestion of just passing in the whole object/resource, and typing as union accordingly.

nawatts commented on src/pages/EnvironmentsPage/environmentsPermissions.ts:
```diff
@@ -0,0 +1,26 @@
+import { IHaveCreator } from 'src/libs/ajax/leonardo/models/core-models';
+import { getTerraUser } from 'src/libs/state';
+
+const makePermissionsProvider = (userEmailGetter: () => string): LeoResourcePermissionsProvider => {
```
This seems like a Terra specific module, but this logic for determining permissions isn't Terra specific. This seems like something that should be handled by environments. And it's annoying that Terra services are so inconsistent... workspaces have permissions returned from the backend but Leo requires duplicating logic on the frontend.

Also, does this need to be a function that returns the user's email instead of the email itself?

msilva-broad replied:
yes, it's an arrow function so that the current value is always retrieved on execution of the permission methods.

The goal is to remove the concern of permissions from the already large Environments component, and also letting AoU determine this logic, and where things like current user email state live, according to their own needs.  Also Justin's comment above highlighted how moving this concern out of the component is a nice overall direction for future needs.

nawatts replied:
Why would the user email change in between invocations of permission functions?

msilva-broad replied:
maybe if they time-out or logout?

nawatts replied:
In which case, they'd be redirected away from this page.

msilva-broad replied:
basically, it's backed by state, which means it can change, so for defensiveness, I made it an arrow function....

msilva-broad committed:
UIE-147 Decouple permissions - Environments page
- PR feedback - tweaked permission provider types to be resource types/unions as appropriate.

evrii reviewed (approved):
This looks like it will be generic enough for use with AoU. It seems like there are a few open threads where people have some differing opinions. I don't have strong enough opinions on any of those threads to be convinced one way or another. Seems like a good step forward. Thanks @msilva-broad !

msilva-broad committed:
UIE-147 Decouple permissions - Environments page
- minor cleanup after rebase
